documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md 

```diff
@@ -0,0 +1,82 @@
+- As an API user, I need to know what exactly this media property is and what data points from the providers are used for it. E.g., does the "creator" refer to the person who created the object or took the picture?
+- As a provider of openly-licensed media, I want to know what media properties Openverse collects so that I can change my API responses.
+
+The information on data types, transformations, and validations will be presented as a table and will be extracted from the code.  Other information about the data selection properties and data discrepancies will be written in a separate markdown file as it is easier to read, write, and lint.
+
+The catalog has a DAG documentation generator that extracts the docstrings from the DAGs and creates a page with the documentation. We can use the same approach to create a documentation page for the media properties. The DAG doc generator runs the code inside docker to extract DAG information. In this project, we can simplify and use the `ast` module to parse static code and extract the docstrings to make the checks faster. This will, however, mean that some of the settings will not be picked up and will have to be updated during the generation. One example of this is the `nullable` field in the `Column` class: it falls back to the value of `not nullable` during the object initiation.
+
+## Outlined Steps
+
+### `just` script
+
+The script parses the SQL DDL files, Python files and markdown files to generate a markdown page with a table and long-form notes. This script can `diff` with the existing documentation page (created in step 4 below) for the `precommit` check and post a note about how to update the docs if differences are detected.
+
+### 1. SQL DDL parsing
+We will use the names of the properties as they are in the upstream database, in the same order as they are in the `image_view`/`audio_view` materialized views. These views are selected over the main `image` and `audio` tables because they are used by the API during data refresh and contain a complete list of properties.
+
+A Python script is written to extract the information from the SQL files into a mapping with the name of the media property and the following values: SQL datatype, and SQL constraints. More data will be added to this mapping in the further steps.
+
+### 2. Parsing of the Python code
+The `Column` class and its child classes describe the validations we use to write the data collected by provider scripts. The `add_item` method has docstrings with short descriptions of what each property is. The information from these items will be added into the mapping from step 1.
+
+### 3. Adding the long-form information from the markdown file
+The markdown file will be parsed and split into the preamble, postamble (???) and the information about each data property. This information would be added to the `description` in the property mapping.
```

Comment:
Which markdown file is this referring to? Also, TIL `postamble` 😄

Reply:
I think I learned it from @AetherUnbound in her DAG documentation generator 😆 

Step 3 refers to the manually created file (I will refer to it as `media_properties.md` in the revised version of this proposal). It would be saved in the `media_properties_gen` folder inside `/catalog/utilities`, in the same way as `dag_doc_gen`. It would have the data that we cannot devise from code or docstrings.
For ease of parsing, we can have a `media_properties.md` for all the properties, and `general.md` for the Preamble and "postamble". The latter would contain notes for inconsistencies that occur on more than one field (like the encoding problem) or anything else that might be useful but can't be put under another heading.

This is a sample of what I think `media_properties` would look like. 

---

# Title
The title of the work. If blank, uses "This work" for the attribution sentence (This work by creator is licensed with CC BY).

## Shape of the data and Selection criteria
We select the default title returned by the provider. It can be blank. Blank values (whether `None` or empty string `""`) are saved as empty string in the database (TODO: check if this is true).

## Existing data problems
- Some media items had incorrectly encoded titles [^1 - Link to a description of Unicode encoding problem in the "postamble"]. This is compensated for in the Frontend (link to the code that fixes title encoding). This problem has been fixed for the items that have been reingested after some time in 2020, but might still persist for items that were not updated since then. Link to issues for fixing the encoding in the catalog/api/frontend.
- Some Wikimedia titles have a shape of "FILE:xxx.svg". The provider script removes them now, but this is still a problem for items that were ingested earlier. The "FILE" and ".extension" are removed in the frontend (link to the code). Link to the issue to fix it in the API.

---

I would like to link all the issues for fixing the data to the documentation, but I'm not sure if adding them under "Existing data problems" is the best solution for that. There would need to be some mechanism of ensuring that the data is updated after the issue is fixed.

Reply:
The idea of separating the pre/postamble into different files is great! That way we don't have to stringify it and reproduce it over and over again with each generation of the file. Do you mind adding all of this description to the IP? It's excellent! 😄